### PR TITLE
Syntax for specifying cumulative inductives

### PR DESCRIPTION
**Pull request by SkySkimmer**
Inductive, Record and synonyms now support syntax `Inductive foo@{=i +j *k l}` to specify variance information for their universes (in Cumulative mode)

Overlays: 
- https://github.com/mit-plv/rewriter/pull/19
- https://github.com/coq-community/paramcoq/pull/54 
- https://github.com/LPCIC/coq-elpi/pull/159 
- https://github.com/Mtac2/Mtac2/pull/291 
- https://github.com/mattam82/Coq-Equations/pull/308
- https://github.com/MetaCoq/metacoq/pull/445

### CODE REVIEW

File: test-suite/success/CumulInd.v
Comment (SkySkimmer) [inline test-suite/success/CumulInd.v:17]: The error message here isn't great due to https://github.com/coq/coq/issues/12661

File: vernac/g_vernac.mlg
Comment (jfehrle) [inline vernac/g_vernac.mlg:382]: How about adding the new/updated productions into the manual with some description?  `inductive_definitions` is shown in the `Inductive` command.  If you like I can push a commit with the updated grammar (or you can figure out how to use doc_grammar to do so yourself with `make doc_gram_rsts`).

Also, this change only applies to `inductive_definition` and not to the 10 other grammar references to `ident_decl`?

**Review by jfehrle — CHANGES_REQUESTED** [review]

File: vernac/g_vernac.mlg
Comment (SkySkimmer) [reply vernac/g_vernac.mlg:382]: > If you like I can push a commit with the updated grammar

Please do.

>Also, this change only applies to inductive_definition and not to the 10 other grammar references to ident_decl?


Correct.

Comment (jfehrle) [reply vernac/g_vernac.mlg:382]: I pushed a commit with the doc/grammar updates, including a few unrelated things from master.  The way the tooling works, updating only part of the syntax is not simple.

**Review by jfehrle — APPROVED** [review]
The grammar in the doc should be fine now.  At some point we need to get [more] explanation of the universe-related syntax.  No opinion on the other parts.

**Comment by SkySkimmer** [comment]
@jfehrle could you redo the docgram update? It got destroyed in the rebase.

**Comment by maximedenes** [comment]
@SkySkimmer can you write a short summary of the change in the PR description? Thanks!

**Comment by SkySkimmer** [comment]
@maximedenes done

File: doc/sphinx/addendum/universe-polymorphism.rst
Comment (gares) [inline doc/sphinx/addendum/universe-polymorphism.rst:263]: I think you should not use "positions" in the following text but rather mention the annotation symbol in the doc.
Eg "a universe `u` has no variance annotation; `*u` requires the variance of u to be irrelevant; `+u` makes u covariant..."

Comment (gares) [reply doc/sphinx/addendum/universe-polymorphism.rst:263]: It would also be nice to have an example/explanation of what variance means (to me it's a very "niche" concept in CS) and possibly a hint to why one may want/need to specify it.

**Comment by SkySkimmer** [comment]
@JasonGross metacoq failed https://gitlab.com/coq/coq/-/jobs/713656562 in make-one-time-file
retry https://gitlab.com/coq/coq/-/jobs/713823030 succeeded
Do you think this will be a recurring problem?

**Comment by jfehrle** [comment]
I pushed a commit with the doc_gram changes.

**Comment by SkySkimmer** [comment]
CI is green, don't believe the github checks

**Comment by SkySkimmer** [comment]
@jfehrle fullgrammar changes have been erased by rebase again

**Comment by JasonGross** [comment]
> @JasonGross metacoq failed https://gitlab.com/coq/coq/-/jobs/713656562 in make-one-time-file
> retry https://gitlab.com/coq/coq/-/jobs/713823030 succeeded
> Do you think this will be a recurring problem?

@SkySkimmer This is due to the interleaving of lines in the build log due to a lack of `--output-sync`:
```
CAMLOPT -c -for-pack Template_coq src/denoter.ml
src/denoter.cmx (real: src/quoter.cmx (real: 0.22, user: 0.240.05, ,u sesry:s :0.16 , 0.02sy,s :m em: 0.0224280,  mkeom:)
34576 ko)
```
How should we fix this?  Is there a way to pass `--output-sync` to all of the invocations of `make` called on the CI?  I can also make the script robust to lines like this.

**Comment by SkySkimmer** [comment]
We could add the option to https://github.com/coq/coq/blob/84d5475169ec0ba3d11928dab11eba6bc3d19d9b/dev/ci/ci-common.sh#L94-L104

**Comment by JasonGross** [comment]
@SkySkimmer Thanks, I've done both that and the robustification in https://github.com/coq/coq/pull/13063

**Comment by jfehrle** [comment]
> fullgrammar changes have been erased by rebase again

I don't understand how a rebase would erase changes.  I pushed a commit updating fullGrammar and orderedGrammar.  Maybe time for you to begin using `make doc_gram_rsts` yourself, described in `doc/README.md`.  If there are no errors and no changes to the .rst files, just commit the result as-is.  And otherwise ask questions if you're uncertain.

**Comment by SkySkimmer** [comment]
Should be ready, hopefully we can merge without another rebase.

File: doc/sphinx/addendum/universe-polymorphism.rst
Comment (gares) [reply doc/sphinx/addendum/universe-polymorphism.rst:263]: What about these comments?

**Comment by gares** [comment]
IMO the doc is still to be improved, see my comments.

File: doc/sphinx/addendum/universe-polymorphism.rst
Comment (SkySkimmer) [reply doc/sphinx/addendum/universe-polymorphism.rst:263]: lost in the github "hidden items"

Comment (SkySkimmer) [reply doc/sphinx/addendum/universe-polymorphism.rst]: Updated to stop using positions

File: doc/sphinx/addendum/universe-polymorphism.rst
Comment (gares) [inline doc/sphinx/addendum/universe-polymorphism.rst:262]: I suggest naming the universes `a b c d` and not `a a' b c`, just because one may think that `'` is a variance modifier.

File: doc/sphinx/addendum/universe-polymorphism.rst
Comment (gares) [inline doc/sphinx/addendum/universe-polymorphism.rst:257]: As I mentioned previously, I don't think `variance` is something any Coq user is expected to know. Even a random CS person may ignore it, it is very specific. Just look at the wikipedia page: https://en.wikipedia.org/wiki/Covariance_and_contravariance_(computer_science)

IMO it should be explained a little here.

**Review by gares — COMMENTED** [review]
IMO, it still needs some doc. It may help to have other opinions/suggestions on this.

File: doc/sphinx/addendum/universe-polymorphism.rst
Comment (SkySkimmer) [reply doc/sphinx/addendum/universe-polymorphism.rst:262]: renamed

File: doc/sphinx/addendum/universe-polymorphism.rst
Comment (SkySkimmer) [reply doc/sphinx/addendum/universe-polymorphism.rst:257]: Variance is already in master, let's not block this PR over it. Feel free to open an issue asking for more doc.

Comment (ppedrot) [reply doc/sphinx/addendum/universe-polymorphism.rst:257]: To prove my point here is a post from yesterday: https://news.ycombinator.com/item?id=24554849.

Comment (SkySkimmer) [reply doc/sphinx/addendum/universe-polymorphism.rst:257]: We should probably have the rules somewhere in the core language section

Comment (Zimmi48) [reply doc/sphinx/addendum/universe-polymorphism.rst:257]: @ppedrot I am siding with @gares here. Not every Coq user is a PL researcher nor an HN reader.

Comment (gares) [reply doc/sphinx/addendum/universe-polymorphism.rst:257]: I don't have strong opinions about the best time to merge this PR, but it looks like it gets you to control a not very documented feature. I see that the manual talks about "universe variance" already, with apparently 0 explanation. I missed that change set, otherwise I would have complained back then.

If you need help to write the doc, we can try to find someone else. I'm not sure I qualify. To me variance is related to subtyping, and Coq's TT does not have subtyping... Oh wait...

Y combinator is name that says it all. @ppedrot go the math department and scream "lambda-calculus!" Then tell me how they look at you.

Comment (Zimmi48) [reply doc/sphinx/addendum/universe-polymorphism.rst:257]: https://coq.inria.fr/refman/language/cic.html#subtyping-rules

Comment (Zimmi48) [reply doc/sphinx/addendum/universe-polymorphism.rst:257]: Note that the word "variance" is not written once on the above page.

Comment (gares) [reply doc/sphinx/addendum/universe-polymorphism.rst:257]: Great, then this is where one should explain how the variance of an inductive in computed and where it plays a role in the rules. I guess it has something to do with `Gamma_P` but I don't really know. I did not even know that conversion could be called a proper subtyping relation (to me it was limited to a "subsort" relation, but I guess that now that it can cross some inductives we can talk about "subtypes").

Comment (Zimmi48) [reply doc/sphinx/addendum/universe-polymorphism.rst:257]: There's also https://coq.inria.fr/refman/language/core/inductive.html#theory-of-inductive-definitions and finally the cumulative inductive types are really only described in https://coq.inria.fr/refman/addendum/universe-polymorphism.html#cumulative-noncumulative. I wish for a consolidation of this material but I'm not sufficiently knowledgeable to really be able to take care of it or even make clear recommendations.

Comment (gares) [reply doc/sphinx/addendum/universe-polymorphism.rst:257]: I'm going trough the doc, and I see that these symbols (+, =, and *) are already mentioned. So I finally agree with @SkySkimmer that it is not for this PR to improve the situation. (but I hope someone does write down how variance is computed somewhere...)

Comment (maximedenes) [reply doc/sphinx/addendum/universe-polymorphism.rst:257]: @gares if it does not already exist, would you mind opening an issue for the documentation of variance?

**Review by maximedenes — APPROVED** [review]
LGTM

**Comment by SkySkimmer** [comment]
CI passes

**Comment by SkySkimmer** [comment]
Elpi failure is spurious, shall we merge? @maximedenes

**Comment by SkySkimmer** [comment]
Since @maximedenes is busy for a few weeks maybe another @coq/kernel-maintainers or @coq/vernac-maintainers can take this?

**Comment by herbelin** [comment]
I never really looked at cumulativity yet, so I would (a priori) let @ppedrot or @mattam82 act.

File: vernac/g_vernac.mlg
Comment (mattam82) [inline vernac/g_vernac.mlg:315]: Maybe we should reconsider "+" for extra declarations and use ".." instead.

File: kernel/inferCumulativity.ml
Comment (mattam82) [inline kernel/inferCumulativity.ml:19]: I think we should enrich BadVariance with the expected and infered variance to provide a more meaningful error message.

**Review by mattam82 — CHANGES_REQUESTED** [review]
The point about the doc is well-taken, I should take care of expanding this part of the refman.
Otherwise the PR looks good. I think the inference code should move outside the kernel and only checking be performed inside but that's for another time. My only request is to have the change on BadVariance implemented so we can get more informative error messages.

File: vernac/g_vernac.mlg
Comment (SkySkimmer) [reply vernac/g_vernac.mlg:315]: >Maybe we should reconsider "+" for extra declarations and use ".." instead.

No, the lookahead works fine so we don't need to bread backward compat.

File: kernel/inferCumulativity.ml
Comment (SkySkimmer) [reply kernel/inferCumulativity.ml]: Updated

**Review by mattam82 — APPROVED** [review]
Nice!

**Comment by SkySkimmer** [comment]
WTF is going on with build:base https://gitlab.com/coq/coq/-/jobs/852148816?
~~~
Fatal error: exception File "typing/env.ml", line 644, characters 26-32: Assertion failed
make[1]: *** [kernel/inferCumulativity.cmo] Error 2
~~~
That's ocamlc crashing.
can't reproduce locally, edge+flambda seems to work https://gitlab.com/coq/coq/-/jobs/852140044
Is this a bug fixed in some ocaml version > base?

I'm assuming there's a bug involving exceptions with inline record so I'll try using a regular exception.

**Comment by SkySkimmer** [comment]
See also https://github.com/ocaml/ocaml/issues/10027

**Comment by mattam82** [comment]
@coqbot merge now

**Comment by mattam82** [comment]
@coqbot merge now

**Pull request merged**
